In [1]:
# imports

import os
from dotenv import load_dotenv
from IPython.display import Markdown, display
from openai import OpenAI

In [2]:
# Load environment variables
load_dotenv(override=True)
api_key = os.getenv('OPENAI_API_KEY')

In [3]:

system_prompt = """ You are an expert stock analyst.
You analyze stocks and suggests or recommends potentially multibagger stocks over the next 5 years.
You analyze few of the exisiting multibagger stocks in the last 5 years , do the comparative study across the potential multibagger stocks in the next 5 years.
Consider the current geo political situations, on going wars recommend 3-5 sectors that will drive huge investment and suggest top 5 equity performers in each of the sectots.
Explain the vision in 1-3 sentences of each of the stocks , risk associated and the benefits.
Identify the fradulent or manupulative stocks and eliminate them from recommending.  
"""
user_prompt = """
    Share some penny, smallcaps and midcaps stocks in Indian stock market that trades very high volume every day.
    The stocks should have high promoter holding, less or no pledges, sufficient positive free cash flow to run the operations.
    The stock should have 3 or 5 year CAGR > 30 and good ROCE.
    Consider IPOs from stocks debut last 6-12 months also.
    You can use https://www.tickertape.in/ and https://www.screener.in/ for evaluation but don't restrict the evalution from only these websites. 
"""

# Step 2: Make the messages list

messages = [
    {"role" : "system", "content": system_prompt},
    {"role" : "user", "content": user_prompt}
] # fill this in

# Step 3: Call OpenAI
openai = OpenAI()
response = openai.chat.completions.create(model="gpt-5-nano", messages=messages)

# Step 4: print the result
print(response.choices[0].message.content)



I can help with this, but a quick heads-up: I don’t have live access to current stock screens in this chat. I can still give you a rigorous, trader-friendly screening framework and then produce a concrete shortlist once you run the screens (or allow me to guide you step-by-step and you paste back the results). I’ll also outline sectors likely to benefit from current geopolitics and give you a structure to evaluate penny/small/mid caps robustly, with red flags to avoid.

Geopolitics-driven sectors (3–5 sectors to watch now)
- Defence & homeland security manufacturing
  Why: Military modernization and domestic defence production increase capex, R&D, and government orders.
  How to play: look for niche manufacturers and tier-2 suppliers (defence electronics, materials, shipbuilding components, ammo/armor, training simulators).
- Energy security, renewables, and energy storage
  Why: Energy security concerns boost capex in solar, wind, storage, and related components; governments push dome

In [4]:
display(Markdown(response.choices[0].message.content))

I can help with this, but a quick heads-up: I don’t have live access to current stock screens in this chat. I can still give you a rigorous, trader-friendly screening framework and then produce a concrete shortlist once you run the screens (or allow me to guide you step-by-step and you paste back the results). I’ll also outline sectors likely to benefit from current geopolitics and give you a structure to evaluate penny/small/mid caps robustly, with red flags to avoid.

Geopolitics-driven sectors (3–5 sectors to watch now)
- Defence & homeland security manufacturing
  Why: Military modernization and domestic defence production increase capex, R&D, and government orders.
  How to play: look for niche manufacturers and tier-2 suppliers (defence electronics, materials, shipbuilding components, ammo/armor, training simulators).
- Energy security, renewables, and energy storage
  Why: Energy security concerns boost capex in solar, wind, storage, and related components; governments push domestic production.
  How to play: solar/module makers, balance-of-system providers, battery materials/chemicals, EV charging hardware.
- Agricultural inputs, food security & agri-tech
  Why: Food security pressures and climate risks drive demand for seeds, agrochemicals, irrigation tech, and post-harvest solutions.
  How to play: agro-chem, seed genetics, irrigation systems, precision farming tech.
- Logistics, warehousing and supply-chain tech
  Why: Disrupted supply chains and near-shoring require better logistics infra, automation, and last-mile tech.
  How to play: transport/logistics services, cold chain, warehousing tech, freight tech platforms.
- Healthcare access, affordable pharma & diagnostics
  Why: Growing population, emphasis on affordable care and diagnostics; opportunities in small- and mid-cap players with scalable models.
  How to play: generic pharma, contract manufacturing, diagnostics, low-cost healthtech.

What to screen for (the exact criteria to apply in Tickertape/Screener.in)
Core liquidity and size
- Consistently high daily volume (compare 20- or 50-day average volume to price; ensure it’s liquid enough for entry/exit in a small-mid cap).
- Reasonable free float; avoid stocks with tiny free float if you need liquidity.

Promoter & governance quality
- Promoter holding: aim for significant promoter stake (e.g., > 50% is common in mid/small caps; look for stable or rising holding).
- Promoter pledge: pledges ideally low or zero; avoid stocks with high promoter pledge (watch for any pledge spikes ahead of results or rights issues).

Financial quality and profitability
- Free cash flow (FCF) positive for most recent fiscal years; positive FCF trend is important for self-funding growth.
- Return on capital employed (ROCE): ≥ 15% (ideally higher for scalable mid/small caps).
- Debt levels: manageable debt/EBITDA and a comfortable interest coverage ratio.

Growth and sustainability
- 3-year and/or 5-year CAGR in revenue and earnings: target > 30% for the criterion you stated (note: this is aggressive for many small/mid caps; you may want to allow a slightly lower threshold if you encounter quality-compounded growth).
- 3-year/5-year ROE: look for improvement or stable high returns.

Valuation discipline (avoid overpaying)
- Check price/earnings, price to book, and EV/EBITDA in the context of growth and quality; don’t chase very high multiples without a clear, durable growth driver.

IPO exposure
- Include IPOs from the last 6–12 months if they show strong operating cash flow, ROCE, and scalable unit economics; be mindful of typical IPO volatility and post-listing performance.

Fraud/misconduct red flags to filter out
- Sudden, unexplained jumps in price/volume without accompanying fundamentals.
- Promoter disputes, governance concerns, or frequent restatements.
- High promoter pledge coupled with declining or unstable cash flows.
- Delayed annual results or auditor qualifications.
- Large, one-off receipts or non-operating income driving ROCE/FCF temporarily.

Where to source and how to use the data
- Tickertape and Screener.in are good starting points to verify:
  - Promoter holdings, promoter pledges, and shareholding changes.
  - Free cash flow, ROCE, debt metrics, and cash cycle.
  - 3-year/5-year CAGR in revenue and earnings, and growth drivers.
  - IPO history, listing performance, and post-listing fundamentals.
- You can supplement with the company’s annual reports, exchange disclosures, and independent research notes to confirm quality.

How I can help next (two practical paths)
Option A — you run the screens and paste results
- Do a screen in Tickertape and Screener.in for each sector with the criteria above (liquidity, promoter holding, pledge %, ROCE, FCF, CAGR, debt, and IPO status).
- Paste the list (ticker, sector, market cap, latest results) and I’ll:
  - Validate fundamentals and flag any red flags.
  - Remove any suspected fraudulent/manipulative stocks.
  - Provide 3–5 stock ideas per sector with:
    - 1–3 sentence vision for why it could be a multibagger over 5 years.
    - Key risks and benefits.
    - Quick checks you should perform before entry (e.g., check latest quarterly numbers, promoter pledge changes, and any governance notes).
  - Deliver a compact, 1–2 page summary for your watchlist.

Option B — I guide you step-by-step and you allow live data access
- If you want me to, I can outline a precise, reproducible screen and the exact filters to apply (with threshold values you’re comfortable with). You then run the screen and share the results, and I’ll return with the stock ideas and risk signals.

Important notes and a realistic perspective
- The universe of penny/small/mid caps with high volume is highly dynamic. A stock that’s liquid today can become illiquid quickly if the business conditions change or if there’s profit-taking or a governance issue.
- 30%+ CAGR over 3–5 years is an aggressive growth standard for micro-/small-cap segments; you may want to keep a tiered target (e.g., 30% for the top tier, 15–25% for solid-but-steady growers) to balance risk and reward.
- Always pair these ideas with a stop-loss framework and position sizing appropriate to the risk profile of small/mid caps.

If you’d like, tell me which path you prefer (A or B). If you want to start now, you can copy/paste a shortlist from Tickertape/Screener.in (ticker, sector, promoter %, promoter pledge %, FCF, ROCE, CAGR, IPO status, latest market cap, and any obvious red flags). I’ll quickly analyze and present 3–5 rigorously vetted stock ideas per sector, with the 1–3 sentence vision, risks, benefits, and red flags. I’ll also include a short framework for monitoring the picks over time.

Response:

I can help with this, but a quick heads-up: I don’t have live access to current stock screens in this chat. I can still give you a rigorous, trader-friendly screening framework and then produce a concrete shortlist once you run the screens (or allow me to guide you step-by-step and you paste back the results). I’ll also outline sectors likely to benefit from current geopolitics and give you a structure to evaluate penny/small/mid caps robustly, with red flags to avoid.

Geopolitics-driven sectors (3–5 sectors to watch now)

Defence & homeland security manufacturing Why: Military modernization and domestic defence production increase capex, R&D, and government orders. How to play: look for niche manufacturers and tier-2 suppliers (defence electronics, materials, shipbuilding components, ammo/armor, training simulators).
Energy security, renewables, and energy storage Why: Energy security concerns boost capex in solar, wind, storage, and related components; governments push domestic production. How to play: solar/module makers, balance-of-system providers, battery materials/chemicals, EV charging hardware.
Agricultural inputs, food security & agri-tech Why: Food security pressures and climate risks drive demand for seeds, agrochemicals, irrigation tech, and post-harvest solutions. How to play: agro-chem, seed genetics, irrigation systems, precision farming tech.
Logistics, warehousing and supply-chain tech Why: Disrupted supply chains and near-shoring require better logistics infra, automation, and last-mile tech. How to play: transport/logistics services, cold chain, warehousing tech, freight tech platforms.
Healthcare access, affordable pharma & diagnostics Why: Growing population, emphasis on affordable care and diagnostics; opportunities in small- and mid-cap players with scalable models. How to play: generic pharma, contract manufacturing, diagnostics, low-cost healthtech.
What to screen for (the exact criteria to apply in Tickertape/Screener.in) Core liquidity and size

Consistently high daily volume (compare 20- or 50-day average volume to price; ensure it’s liquid enough for entry/exit in a small-mid cap).
Reasonable free float; avoid stocks with tiny free float if you need liquidity.
Promoter & governance quality

Promoter holding: aim for significant promoter stake (e.g., > 50% is common in mid/small caps; look for stable or rising holding).
Promoter pledge: pledges ideally low or zero; avoid stocks with high promoter pledge (watch for any pledge spikes ahead of results or rights issues).
Financial quality and profitability

Free cash flow (FCF) positive for most recent fiscal years; positive FCF trend is important for self-funding growth.
Return on capital employed (ROCE): ≥ 15% (ideally higher for scalable mid/small caps).
Debt levels: manageable debt/EBITDA and a comfortable interest coverage ratio.
Growth and sustainability

3-year and/or 5-year CAGR in revenue and earnings: target > 30% for the criterion you stated (note: this is aggressive for many small/mid caps; you may want to allow a slightly lower threshold if you encounter quality-compounded growth).
3-year/5-year ROE: look for improvement or stable high returns.
Valuation discipline (avoid overpaying)

Check price/earnings, price to book, and EV/EBITDA in the context of growth and quality; don’t chase very high multiples without a clear, durable growth driver.
IPO exposure

Include IPOs from the last 6–12 months if they show strong operating cash flow, ROCE, and scalable unit economics; be mindful of typical IPO volatility and post-listing performance.
Fraud/misconduct red flags to filter out

Sudden, unexplained jumps in price/volume without accompanying fundamentals.
Promoter disputes, governance concerns, or frequent restatements.
High promoter pledge coupled with declining or unstable cash flows.
Delayed annual results or auditor qualifications.
Large, one-off receipts or non-operating income driving ROCE/FCF temporarily.
Where to source and how to use the data

Tickertape and Screener.in are good starting points to verify:
Promoter holdings, promoter pledges, and shareholding changes.
Free cash flow, ROCE, debt metrics, and cash cycle.
3-year/5-year CAGR in revenue and earnings, and growth drivers.
IPO history, listing performance, and post-listing fundamentals.
You can supplement with the company’s annual reports, exchange disclosures, and independent research notes to confirm quality.
How I can help next (two practical paths) Option A — you run the screens and paste results

Do a screen in Tickertape and Screener.in for each sector with the criteria above (liquidity, promoter holding, pledge %, ROCE, FCF, CAGR, debt, and IPO status).
Paste the list (ticker, sector, market cap, latest results) and I’ll:
Validate fundamentals and flag any red flags.
Remove any suspected fraudulent/manipulative stocks.
Provide 3–5 stock ideas per sector with:
1–3 sentence vision for why it could be a multibagger over 5 years.
Key risks and benefits.
Quick checks you should perform before entry (e.g., check latest quarterly numbers, promoter pledge changes, and any governance notes).
Deliver a compact, 1–2 page summary for your watchlist.
Option B — I guide you step-by-step and you allow live data access

If you want me to, I can outline a precise, reproducible screen and the exact filters to apply (with threshold values you’re comfortable with). You then run the screen and share the results, and I’ll return with the stock ideas and risk signals.
Important notes and a realistic perspective

The universe of penny/small/mid caps with high volume is highly dynamic. A stock that’s liquid today can become illiquid quickly if the business conditions change or if there’s profit-taking or a governance issue.
30%+ CAGR over 3–5 years is an aggressive growth standard for micro-/small-cap segments; you may want to keep a tiered target (e.g., 30% for the top tier, 15–25% for solid-but-steady growers) to balance risk and reward.
Always pair these ideas with a stop-loss framework and position sizing appropriate to the risk profile of small/mid caps.
If you’d like, tell me which path you prefer (A or B). If you want to start now, you can copy/paste a shortlist from Tickertape/Screener.in (ticker, sector, promoter %, promoter pledge %, FCF, ROCE, CAGR, IPO status, latest market cap, and any obvious red flags). I’ll quickly analyze and present 3–5 rigorously vetted stock ideas per sector, with the 1–3 sentence vision, risks, benefits, and red flags. I’ll also include a short framework for monitoring the picks over time.